### ISSUE TITLE
retire_users management command doesn't unenroll learner from the course in edX

### ISSUE THREAD

**Issue by briangrossman** [mitodl/mitxonline#697]
### Steps to Reproduce

- Find a learner who is enrolled in at least one course
- Run `retire_users` command to retire the learner

### Expected Behavior

- The learner should be unenrolled from the courses they're enrolled in in edX

### Actual Behavior

- The learner's enrollment is removed from the MITx Online Django admin
- ... but the learner is still actively enrolled in the course on the edX side

### Screenshot or Screencast

Enrollments before retiring learner (Django admin):
![Before](https://user-images.githubusercontent.com/21342568/179052516-05687791-faa2-4236-ac06-d34900126eef.png)

Retire Learner:
![image](https://user-images.githubusercontent.com/21342568/179052570-3cdb69da-2e2a-417c-b7d0-3d27ae1ff178.png)

Enrollments after retiring learner (Django admin): 
![After](https://user-images.githubusercontent.com/21342568/179052602-6dc124d9-6668-42ec-b035-c667e33e00f3.png)

Enrollment on the edX side:
![image](https://user-images.githubusercontent.com/21342568/179052824-870636dd-edee-4876-abd9-0a552c2c1fbd.png)

### related issue:

https://github.com/mitodl/ol-infrastructure/issues/1007

**Comment by jkachel** [comment]
Related Sentry log message?
https://sentry.io/organizations/mit-office-of-digital-learning/issues/3426657144/?project=5864687&referrer=slack

**Comment by briangrossman** [comment]
Kinda, but not really... I accidentally tried to remove my test user from production first--they were on RC. This error was generated when I accidentally ran the retire command on production. So, it's a red herring. 

Here's the code for the management command, BTW: https://github.com/mitodl/mitxonline/blob/main/users/management/commands/retire_users.py

**Comment by pdpinch** [comment]
While we're investigating this, can we also revisit whether there is a proper API in open edX for retiring users that we can use?

**Comment by pdpinch** [comment]
Take a look at this API: https://github.com/openedx/edx-platform/blob/7686014144930bb9fda4ae8beb66f0392855e55a/lms/djangoapps/bulk_user_retirement/urls.py#L11-L15

**Comment by pdpinch** [comment]
That API may not work without some further configuration of open edX. When Tobiase tried to retire a user in production, we got the error: 

https://sentry.io/organizations/mit-office-of-digital-learning/issues/3513101534/?project=5939801&query=is%3Aunresolved

**Comment by asadali145** [comment]
API mentioned above is a bulk retirement API, Needs the `ENABLE_BULK_USER_RETIREMENT` FEATURE flag enabled.
Also, we have another API to retire a single user by username, but I believe the bulk retirement would be better. https://github.com/openedx/edx-platform/blob/7686014144930bb9fda4ae8beb66f0392855e55a/openedx/core/djangoapps/user_api/urls.py#L167


Retirement in Open edX needs `Retirement States` configured either from Django Admin or by a management command which uses settings of `Retirement States`.

**Comment by asadali145** [comment]
@pdpinch Let us know if we can go with the first option. It will require some configs and an automated pipeline to process the retirements. Here are some of the findings:

We have 2 options to retire/unenroll learners on edX:

1. We can configure proper user retirement in edX. 
   -  This includes setting up the `RetirementStates` using https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/user_retire/service_setup.html#retirement-states. Minimal states required are:
    ```
    RETIREMENT_STATES = [
        'PENDING',
    
        'RETIRING_ENROLLMENTS',
        'ENROLLMENTS_COMPLETE',
    
        'RETIRING_LMS_MISC',
        'LMS_MISC_COMPLETE',
    
        'RETIRING_LMS',
        'LMS_COMPLETE',
    
        'ERRORED',
        'ABORTED',
        'COMPLETE',
    ]
    ```
   - Next step is to create a service user for retirement using https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/user_retire/service_setup.html#retirement-service-user. Client ID and Secret will be used in the retirement pipeline config(discussed later). The username of the service user will be added to the edX settings against `RETIREMENT_SERVICE_WORKER_USERNAME`.
   - Next thing is setting up an automated pipeline that uses https://github.com/edx/tubular. edX has provided scripts for retirement in https://github.com/edx/tubular. Jenkins implementation of user retirement is private, so, we will have to create a pipeline.
   - Tubular requires a config file in YAML format. In our case, It can be like this:
    ```
    client_id: <service_user_client_id>
    client_secret: <service_user_client_secret>
    
    base_urls:
            lms: <lms_base_url>
    
    retirement_pipeline:
            - ['RETIRING_ENROLLMENTS', 'ENROLLMENTS_COMPLETE', 'LMS', 'retirement_unenroll']
            - ['RETIRING_LMS_MISC', 'LMS_MISC_COMPLETE', 'LMS', 'retirement_lms_retire_misc']
            - ['RETIRING_LMS', 'LMS_COMPLETE', 'LMS', 'retirement_lms_retire']
    
    ```
   - Details about setting up the user retirement driver scripts can be found in https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/user_retire/driver_setup.html#setting-up-the-user-retirement-driver-scripts.

2. Other option is to hit the APIs to retire the user and unenroll from all the courses. Currently, exploring this option as we are facing authentication issues with the related APIs.

**Comment by asadali145** [comment]
We were trying to go with the 2nd option as it requires fewer configs and less management. But we are still stuck with the APIs.

**Comment by asadali145** [comment]
The best possible approach after R&D is as below.

- It seems like our DevOps team will have to set up this pipeline provided by edX https://github.com/openedx/configuration/tree/master/playbooks/roles/user_retirement_pipeline.
- We will have to update the minimum required `retirement states` and the pipeline according to the states.
- Once all of this is set up, We can run the `retire_users.sh` script on demand.

CC: @pdpinch @arslanashraf7

**Comment by pdpinch** [comment]
- [ ] @mitodl/devops can you take a look at https://github.com/openedx/configuration/tree/master/playbooks/roles/user_retirement_pipeline ? Is concourse the right tool for this? 

- [ ] Can you also advise @asadali145 on where to add the configuration settings? (salt-ops, ol-infrastructure?) 

- [ ] @asadali145 question: What's the interface for running `retire_users.sh` script?

**Comment by asadali145** [comment]
> @asadali145 question: What's the interface for running retire_users.sh script?
I don't think there is any interface for this one but we have 2 options, Either schedule this to run periodically or run this manually from shell maybe. Not sure how we are going to set this up with our infrastructure.

**Comment by asadali145** [comment]
@mitodl/devops Any update on this one?

**Comment by asadali145** [comment]
Just to be transparent and FYI @pdpinch, I have never worked on ol-infrastructure and I also do not have any kind of domain knowledge of ol-infrastructure. 

For me to work on this, I will have to get some domain knowledge and will have to learn and implement things that might require someone to help me understand the architecture. I will ask @arslanashraf7 to help me once he is back from leaves and for further help, I would ask the DevOps team to help me out on this one. Currently, I just don't know where to start.

I'd suggest that maybe someone from DevOps should take over this ticket and I can help him understand the things that we need to set up.

**Comment by pdpinch** [comment]
@blarghmatey I think you're going to have to give Asad some more direction about how to add the settings to ol-infrastructure. It's not obvious to me 1) which file the settings go into, 2) where in the file they go and 3) how they should be formatted.

**Comment by pdpinch** [comment]
When we get this working, there are two accounts that shoudl be retired, before this is closed

- [ ] https://odl.zendesk.com/agent/tickets/134042
- [ ] https://odl.zendesk.com/agent/tickets/131058

**Comment by pdpinch** [comment]
@blarghmatey now that https://github.com/mitodl/ol-infrastructure/pull/1118 is merged, what are the next steps? I think this is correct: 

- [ ] Test the retirement shell command on mitxonline QA
- [ ] deploy the retirement settings to mitxonline production 
- [ ] retire the users mentioned in the tickets listed above (this list may have grown since I made that comment)
- [ ] update the mitxonline management command to replicate the steps of the open edX retirement shell script. We need a list of these steps.

**Comment by blarghmatey** [comment]
I just ran the command on the QA environment:
```
edxapp@ip-10-21-1-248:~$ python edx-platform/manage.py lms retire_user --user_email pdpinch+retireme@hawklake.com --username pdpinch-retireme
2022-10-12 13:52:29,184 INFO 1881 [openedx.core.djangoapps.user_api.management.commands.retire_user] [user None] [ip None] retire_user.py:100 - User succesfully moved to the retirment pipeline
```

**Comment by pdpinch** [comment]
Looks like it worked on QA: https://courses-qa.mitxonline.mit.edu/admin/auth/user/302

Did this go out with the production release yesterday?

**Comment by blarghmatey** [comment]
Yes, that configuration is in production as of yesterday. From reading through the retirement workflow docs, I don't think that the user in QA has actually reached an end status in the state machine. I'll take another look at that today.

**Comment by blarghmatey** [comment]
Confirming my suspicion: https://courses-qa.mitxonline.mit.edu/admin/user_api/userretirementstatus/1/change/
That user is in a `PENDING` state currently until we run the user retirement code again (after the proper "cooldown days" period) and propagate them to the next step.

**Comment by pdpinch** [comment]
@blarghmatey regarding this todo item:

> update the mitxonline management command to replicate the steps of the open edX retirement shell script. We need a list of these steps.

Are these the steps? 

1.   'RETIRING_ENROLLMENTS',
2.   'RETIRING_LMS_MISC',
3.   'RETIRING_LMS',

What is the API end-point/docs that we need to use for this? Can you give @asadali145 some pointers?

**Comment by asadali145** [comment]
> python edx-platform/manage.py lms retire_user --user_email pdpinch+retireme@hawklake.com --username pdpinch-retireme

@pdpinch We need to run tubular scripts to complete the retirement.

Running the above command moves a user to the retirement pipeline. As in this case, the User is moved to the `PENDING` retirement state.

Now, to complete this retirement process, we need to use https://github.com/openedx/tubular. Tubular requires a config file in YAML. Tubular provides driver scripts for this purpose. Steps are mentioned [here](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/user_retire/driver_setup.html#setting-up-the-user-retirement-driver-scripts) in edX guide. 

In short, once tubular is set up with the config file. We can do:

1. `scripts/retire_one_learner.py --config_file=path/to/config.yml --username=<username-of-learner-to-retire>`

**Comment by pdpinch** [comment]
According to @blarghmatey , there is a REST API or APIs that we can access as well. I would very much prefer that this process can be managed from mitxonline, through a management command.

**Comment by arslanashraf7** [comment]
Looking at the communication above and based on the possibilities for making this work right from MITxOnline or any other app for that matter I was thinking of creating some kind of plugin for this. We've already been following the practice of creating our own custom plugins for edX in [open-edx-plugins](https://github.com/mitodl/open-edx-plugins).

An example that might be relatable in this scenario would be our [ol_openedx_course_export](https://github.com/mitodl/open-edx-plugins/tree/main/src/ol_openedx_course_export) where we created our own API to export the course to S3 with just a course key in the POST call.

If it makes sense, I think it's worth putting some effort into checking the feasibility of this use case following the [Open edX plugins Architecture](https://github.com/openedx/edx-django-utils/tree/master/edx_django_utils/plugins).

We can simply create an API or maybe a management command through this plugin and respectively call the API from anywhere or command from edX to perform all the needed steps for a user's retirement. 

FYI, @asadali145 

Let's hear thoughts from @pdpinch and @blarghmatey.

**Comment by pdpinch** [comment]
> Confirming my suspicion: https://courses-qa.mitxonline.mit.edu/admin/user_api/userretirementstatus/1/change/
That user is in a PENDING state currently until we run the user retirement code again (after the proper "cooldown days" period) and propagate them to the next step.

@blarghmatey I've been cooling down for 3 weeks... what do we need to do here?

**Comment by blarghmatey** [comment]
I think the next step is to run the management command again. I can do that tomorrow.

**Comment by asadali145** [comment]
Hi, I think we won't be able to complete the retirement without the tubular scripts. There is no command in MITxOnline or in edX to do that for us. Cooldown time can be zero as well while retiring with tubular. 

I am setting things up on my new machine to create a plugin for retirement. Will start my work on that ASAP.

**Comment by asadali145** [comment]
Leaving our thoughts on current implementation that we are working on:

We considered the following options:

- Create a package using [ol-django](https://github.com/mitodl/ol-django) and use [tubular](https://github.com/openedx/tubular) as its dependency in our apps. It turned out that we cannot do this as _Tubular is not compatible with Python 3.9+_.
- Create a package for user retirement in https://github.com/mitodl/open-edx-plugins and use [tubular](https://github.com/openedx/tubular) as its dependency. It turned out that _we cannot install tubular in edX instance due to dependency version conflicts_.
- Last option was use code from tubular and create our own user retirement plugin in [ol-django](https://github.com/mitodl/ol-django). Currently, we are working on creating a package in [ol-django](https://github.com/mitodl/ol-django).

**Comment by asadali145** [comment]
Config for retirement Pipeline.
```
client_id: <retirement service app client ID>
client_secret: <retirement service app secret>
base_urls:
    lms: http://localhost:18000/
retirement_pipeline:
    - ['RETIRING_FORUMS', 'FORUMS_COMPLETE', 'LMS', 'retirement_retire_forum']
    - ['RETIRING_ENROLLMENTS', 'ENROLLMENTS_COMPLETE', 'LMS', 'retirement_unenroll']
    - ['RETIRING_NOTES', 'NOTES_COMPLETE', 'LMS', 'retirement_retire_notes']
    - ['RETIRING_PROCTORING', 'PROCTORING_COMPLETE', 'LMS', 'retirement_retire_proctoring_data']
    - ['RETIRING_LMS_MISC', 'LMS_MISC_COMPLETE', 'LMS', 'retirement_lms_retire_misc']
    - ['RETIRING_LMS', 'LMS_COMPLETE', 'LMS', 'retirement_lms_retire']

```

**Comment by pdpinch** [comment]
@feoh this is the issue that started it all for why we need to run [open edX Tubular](https://github.com/openedx/tubular)

**Comment by feoh** [comment]
I’ve written an [RFC](https://github.com/mitodl/hq/discussions/1741) of the current process, and how it might be improved.

**Comment by pdpinch** [comment]
Thanks @feoh 

I'm moving this back into Ready and removing your assignment so one of the software engineers can propose an update to the management command and/or process so we don't need so many manual steps.

**Comment by asadali145** [comment]
Leaving my thoughts on automating manual steps:

- Step 0: Interface for initiating retirement.
  - Provide an edX like account deletion interface to the users in our apps. EdX has provided it on the `Account` page. OR
  - Create an interface for the admins to retire users through Django Admin. OR
  - Update/Create retirement management commands in our apps to handle the edX retirement initiation.
- Step 1: Moving a user to the retirement pipeline i.e. Create retirement in Pending State. Currently, we need to run a management command in edX shell to initiate retirement process.
  - Once we have an interface for the user retirement in our apps, we can call the edX bulk retirement API to initiate the retirement. API will require admin permissions. OR
  - We can also create an edX plugin for this purpose but it will complicate a few things. It can be an edX plugin to initiate the retirement.
- Step 3: Processing the user retirement with Tubular/Concourse pipeline.
  - I am not really sure about this but we can schedule it run once or twice a day to process the retirements.


@pdpinch @arslanashraf7 @feoh What do you think about this?

**Comment by pdpinch** [comment]
> Update/Create retirement management commands in our apps to handle the edX retirement initiation.

We have a management command already in the mitxonline app, but it doesn't currently include your "Step 1"

https://github.com/mitodl/mitxonline/blob/7070d812c71798f2bd9fd0257b6dd5a98c179256/users/management/commands/retire_users.py#L33

> we can call the edX bulk retirement API to initiate the retirement. API will require admin permissions.

Can you link to some documentation for this API? 

> we can schedule it run once or twice a day to process the retirements.

@feoh Is there an API to trigger the pipeline, or am I asking for too much? If not, I think a regular schedule would be OK unless you have a concern.

**Comment by asadali145** [comment]
> We have a management command already in the mitxonline app, but it doesn't currently include your "Step 1"

Yep, we can update the command to initiate retirement using an admin user. Still, it would require us to access the shell to run the command, I am not sure how we run commands on RC and Prod. When I suggested providing an edX-like interface to the learners, I had self-service in my mind referring to your [comment](https://github.com/mitodl/hq/discussions/1741#discussioncomment-6422044) `If we find a way to automate all these steps, we could make it self-service.`

> Can you link to some documentation for this API?

Here is the [API](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/bulk_user_retirement/views.py#L18). I am not able to find any other docs, API docs are good enough I think.

**Comment by feoh** [comment]
@pdpinch It already runs on a [schedule](https://github.com/mitodl/ol-infrastructure/blob/72a7506cb9b8c0b23f8e26c91bc2fffadc938042/src/ol_concourse/pipelines/open_edx/tubular/tubular.py#L29). Tobias suggested weekly so that's what it currently runs at. We can run it more frequently if you think that makes more sense.

As to an API for kicking it off, @blarghmatey I don't know if we have any prior art on this. Could you please chime in if so?

If not I found [this StackOverflow entry](https://stackoverflow.com/questions/44473538/rest-api-to-trigger-a-concourse-pipeline-job) and it doesn't seem overly onerous although I don't know what the auth implications are in our environment.

**Comment by pdpinch** [comment]
@asadali145 can we add the `bulk_user_retirement` api to our [edx-api-client library](https://github.com/mitodl/edx-api-client). I believe the mitxonline application is already configured with an admin user, so it can use the force flag on the enrollments API.

**Comment by asadali145** [comment]
I have created the draft PRs in MITxOnline and edx-api-client but we cannot use these changes immediately as there is an issue in edX bulk retirement API.

Currently, the URL of the API is [`v1/accounts/bulk_retire_users`](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/bulk_user_retirement/urls.py#L12C10-L12C39). The API uses [JWT Authentication](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/bulk_user_retirement/views.py#L37). When we try to access the API at `v1/accounts/bulk_retire_users` it always redirects to the login page and we are not able to authenticate the API. **But if we update the edX URL to `api/v1/accounts/bulk_retire_users`, everything works fine.**

We are thinking about creating an edX discussion post around it.

@pdpinch What are you thoughts on this one?

**Comment by blarghmatey** [comment]
The redirect is likely because of the behavior of the `social-auth-mitxpro` package that we use. There is a setting to add an allowlist for paths that won't be forced into a redirect. https://github.com/mitodl/ol-infrastructure/blob/main/src/bilder/images/edxapp/templates/edxapp/mitxonline/common_values.yml.tmpl#L360
